schema/Metrics.v1.0.0.schema.json

```diff
@@ -0,0 +1,267 @@
+          "description": "The value to aggregate."
+        }
+      },
+      "description": "The definition of an Average metric. This metric kind measures the average of an event property."
+    },
+    "ObservedEvent": {
+      "type": "object",
+      "description": "An event observed by a metric.",
+      "properties": {
+        "eventName": {
+          "type": "string",
+          "description": "The name of the event.",
+          "minLength": 1,
+          "maxLength": 200
+        },
+        "filter": {
+          "$ref": "#/definitions/EventCondition",
+          "description": "[Optional] A condition to filter events."
+        }
+      },
+      "required": ["eventName"]
+    },
+    "EventCondition": {
+      "type": "string",
+      "description": "A condition to filter events. Accepts a Kusto Query Language (KQL) filter predicate. \nAllowed expressions evaluate to a bool data type and use a subset of KQL syntax:\nEvent properties: plain or bracket notation.\nLiterals: bool, long, real, string.\nComparison operators: ==, !=, <, <=, >, >=.\nBoolean operators: and, or.\nGrouping operations: parentheses.",
+      "minLength": 1,
+      "maxLength": 1000
+    },
+    "AggregatedValue": {
+      "type": "object",
+      "required": ["eventName", "eventProperty"],
+      "properties": {
+        "eventName": {
+          "type": "string",
+          "minLength": 1,
+          "maxLength": 200,
+          "description": "The name of the event."
+        },
+        "filter": {
+          "$ref": "#/definitions/EventCondition",
+          "description": "[Optional] A condition to filter events."
+        },
+        "eventProperty": {
+          "type": "string",
+          "minLength": 1,
+          "maxLength": 200,
+          "pattern": "^[a-zA-Z0-9_ -.]+$",
+          "description": "The key of the event property to aggregate."
+        }
+      },
+      "description": "An event property value aggregated by a metric."
+    },
+    "MetricDefinition": {
+      "description": "The metric definition, which determines how the metric value is calculated from event data.",
+      "type": "object",
+      "properties": {
+        "kind": {
+          "type": "string",
+          "enum": [
+            "EventCount",
+            "UserCount",
+            "EventRate",
+            "UserRate",
+            "Sum",
+            "Average",
+            "Percentile"
+          ],
+          "description": "Discriminator property for MetricDefinition."
+        }
+      },
+      "anyOf": [
+        { "$ref": "#/definitions/EventCountDefinition" },
+        { "$ref": "#/definitions/UserCountDefinition" },
+        { "$ref": "#/definitions/EventRateDefinition" },
+        { "$ref": "#/definitions/UserRateDefinition" },
+        { "$ref": "#/definitions/SumDefinition" },
+        { "$ref": "#/definitions/AverageDefinition" },
+        { "$ref": "#/definitions/PercentileDefinition" }
+      ],
+      "required": ["kind"]
```

Comment:
Does this mean the customer can specify "kind" without any of the other fields in "definition"? Can we add a test case that checks this is invalid:
```json
"definition": {"kind": "EventCount"}
```

Reply:
Added a test to validate this scenario that user can't provide definition with just kind property. It has to be one of the valid definition in the oneof list.